contracts/auctions/FungibleAuction.sol

```diff
@@ -0,0 +1,165 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.7.3;
+
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+
+import "../IAuctionHouse.sol";
+import "./IFungibleAuction.sol";
+
+/// @title Auction for fungible tokens
+contract FungibleAuction is IFungibleAuction {
+    using SafeERC20 for IERC20;
+    using SafeMath for uint256;
+
+    /// @dev Only set if auction was created via the auction house factory. Used for callbacks.
+    IAuctionHouse public auctionHouse;
+
+    /// @notice Address of the contract governing the token that is being auctioned.
+    IERC20 public immutable sellTokenAddress;
+
+    /// @notice Address of the contract governing the token used to buy in this auction.
+    IERC20 public immutable buyTokenAddress;
+
+    /// @notice Creator of the auction and address auctioning their tokens.
+    address public immutable auctioneer;
+
+    /// @notice Parameters of this specific auction.
+    uint256 public immutable initialAmount;
+    uint256 public immutable startTimestamp;
+    uint256 public immutable duration;
+    uint256 public immutable maxPrice;
+    uint256 public immutable minPrice;
+
+    /// @notice The last paid price, for Dutch Auctions this is also the lowest price.
+    uint256 public lowestSellPrice;
+
+    constructor(
+        IERC20 sellTokenAddress_,
+        IERC20 buyTokenAddress_,
+        uint256 initialAmount_,
+        address auctioneer_,
+        uint256 startTimestamp_,
+        uint256 duration_,
+        uint256 maxPrice_,
+        uint256 minPrice_,
+        bool isAuctionHouse
+    ) {
+        require(maxPrice_ >= minPrice_, "Auction: Price must go down");
+        require(startTimestamp_ + duration_ >= block.timestamp, "Auction: Cannot end in the past");
+
+        if (isAuctionHouse) {
+            auctionHouse = IAuctionHouse(msg.sender);
+        }
+
+        initialAmount = initialAmount_;
+        auctioneer = auctioneer_;
+        sellTokenAddress = sellTokenAddress_;
+        buyTokenAddress = buyTokenAddress_;
+        startTimestamp = startTimestamp_;
+        duration = duration_;
+        maxPrice = maxPrice_;
+        minPrice = minPrice_;
+    }
+
+    /// @dev Buy `amount` tokens. Will buy whatever is left if `amount` is not available in its entirety.
+    function buy(uint256 amount) external override {
+        require(isActive(), "Auction: not active");
+
+        uint256 availableAmount = sellTokenAddress.balanceOf(address(this));
+
+        require(availableAmount > 0, "Auction: sold out");
+
+        // Allow buyer to buy whatever is left.
+        if (amount > availableAmount) {
+            amount = availableAmount;
+        }
+
+        // The number of tokens required to buy the requested amount (or what is left of it).
+        uint256 buyPrice = getPriceInBuyTokens(amount);
```

Comment:
nit: How about renaming getPrice to getRate, and inlining the multiplication of rate and amount here instead?

Reply:
"Rate", that's the word I was looking for! Will change!